vllm/entrypoints/openai/tool_parsers.py

```diff
@@ -0,0 +1,624 @@
+                        delta = None
+                    elif not cur_arguments and prev_arguments:
+                        logger.error('INVARIANT - impossible to have arguments reset mid-arguments')
+                        delta = None
+                    elif cur_arguments and not prev_arguments:
+                        cur_arguments_json = json.dumps(cur_arguments)
+                        logger.info(f'Finding {new_text} in |{cur_arguments_json}|')
+                        arguments_delta = cur_arguments_json[:cur_arguments_json.index(new_text) + len(new_text)]
+                        logger.info(f'First tokens in arguments received: {arguments_delta}')
+                        delta = DeltaMessage(tool_calls=[
+                            DeltaToolCall(index=self.current_tool_id, function=DeltaFunctionCall(
+                                arguments=arguments_delta
+                            ).model_dump(exclude_none=True))
+                        ])
+                        self.streamed_args_for_tool[self.current_tool_id] += arguments_delta
+
+                    elif cur_arguments and prev_arguments:
+                        cur_args_json = json.dumps(cur_arguments)
+                        prev_args_json = json.dumps(prev_arguments)
+                        logger.info(f'Searching for diff between \n{cur_args_json}\n{prev_args_json}')
+                        argument_diff = extract_intermediate_diff(cur_args_json, prev_args_json)
+                        logger.info(f'got arguments diff: {argument_diff}')
+                        delta = DeltaMessage(tool_calls=[
+                            DeltaToolCall(index=self.current_tool_id, function=DeltaFunctionCall(
+                                arguments=argument_diff
+                            ).model_dump(exclude_none=True))
+                        ])
+                        self.streamed_args_for_tool[self.current_tool_id] += argument_diff
+                    else:
+                        # try parsing it with regular JSON - if it works we're at the end, and we need to send the
+                        #   difference between tokens streamed so far and the valid JSON
+                        delta = None
+
+                # check to see if the name is defined and has been sent. if so, stream the name - otherwise keep waiting
+                # finish by setting old and returning None as base case
+                self.prev_tool_call_arr = tool_call_arr
+                return delta
+
+            except Exception as e:
+                logger.error(f'Error trying to handle streaming tool call: {e}')
+                logger.info(f'Skipping chunk as a result of tool streaming extraction error')
+                return None
+
+
+class Hermes2ProToolParser(ToolParser):
+    tool_call_start_token: str = '<tool_call>'
+    tool_call_end_token: str = '</tool_call>'
+    tool_call_start_token_id: int = 128004
+    tool_call_end_token_id: int = 128011
```

Comment:
Hi! Would it be possible to avoid hardcoding the integer token ids? For example, use `tool_call_start_token_id = tokenizer.vocab['<tool_call>']` and `tool_call_end_token_id = tokenizer.vocab['</tool_call>']` instead of hardcoded numbers such as 128004 and 128011. This would allow models fine-tuned with different tokenizers to use the parser.

Reply:
great idea!

Reply:
558461a7